File: src/frontend/src/api/groups.ts
import { api } from './client'
import { sessionsApi } from './sessions'
import type { ApiGroup, Session } from '../types'

export interface CreateGroupInput {
  name: string
  description?: string
  /** 初始成员 Agent id 列表；协调者由后端自动加入，不在此列。 */
  member_ids?: string[]
  /** 群组工作目录（必填）。 */
  workdir: string
}

export interface NameCheckResult {
  available: boolean
  reason?: string
}

export const groupsApi = {
  list: () => api.get<ApiGroup[]>('/api/groups'),
  create: (input: CreateGroupInput) => api.post<ApiGroup>('/api/groups', input),
  rename: (id: string, name: string) => api.patch<ApiGroup>(`/api/groups/${id}`, { name }),
  remove: (id: string) => api.del<void>(`/api/groups/${id}`),
  checkName: (name: string) =>
    api.get<NameCheckResult>(`/api/groups/check-name?name=${encodeURIComponent(name)}`),

  /**
   * 找回已有群聊 Session 或创建新的。
   * 当前后端 GET /api/sessions 不支持 ?group_id= 过滤，先 list(type=group) 再筛选。
   * 后端加上 group_id query 后可简化为单次 GET。
   */
  findOrCreateSession: async (groupId: string): Promise<Session> => {
    const all = await sessionsApi.list({ type: 'group' })
    const hit = all.find((s) => s.group_id === groupId && s.type === 'group')
    if (hit) return hit
    return sessionsApi.createGroup(groupId)
  },

  /**
   * t7 拓展：群组置顶/取消置顶。
   *
   * 设计权衡：复用 backing Session.pinned 而非给 Group 实体加列
   * - 收益：0 alembic 迁移 + 0 entity 改动 + 复用 t7 已落地的 Session.pinned 列 + service + 401/403/422 校验
   * - 代价：删 group 时级联删 session → pin 状态丢失；新建 group 自动建 session → 默认 pinned=false
   * - 决策点：等真正出现"删 group 要保留 pin"需求时再迁到 Group.pinned（alembic 0024 + entity）
   *
   * 实现：没 sessionId 时先 findOrCreateSession 拿，再 PATCH `/api/sessions/{id}` body={pinned}
   * 复用 sessionsApi.patch → 401/403/422 链路与单聊一致
   */
  togglePin: async (
    groupId: string,
    sessionId: string | null | undefined,
    nextPinned: boolean,
  ): Promise<void> => {
    let sid = sessionId
    if (!sid) {
      const s = await groupsApi.findOrCreateSession(groupId)
      sid = s.id
    }
    await sessionsApi.patch(sid, { pinned: nextPinned })
  },
}

File: src/backend/app/api/routers/groups.py
"""群组路由（设计文档 §四）。创建群组自动生成协调者 + 写成员。"""

from __future__ import annotations

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, status

from app.api.deps import get_group_service
from app.application.commands import (
    CreateGroupCommand,
    DeleteGroupCommand,
    RenameGroupCommand,
)
from app.application.dto import GroupResponse
from app.application.services import GroupService
from app.infrastructure.db.base import session_factory
from app.infrastructure.db.models import SessionModel
from app.schemas.group import (
    GroupCoordinatorOut,
    GroupCreateRequest,
    GroupMemberOut,
    GroupOut,
    GroupRenameRequest,
    NameCheckOut,
)
from sqlalchemy import select

router = APIRouter(prefix="/api/groups", tags=["groups"])

ServiceDep = Annotated[GroupService, Depends(get_group_service)]


def _to_out(
    resp: GroupResponse,
    pinned: bool = False,
    session_id: UUID | None = None,
) -> GroupOut:
    return GroupOut(
        id=resp.id,
        name=resp.name,
        description=resp.description,
        coordinator=GroupCoordinatorOut(**resp.coordinator.__dict__),
        members=[GroupMemberOut(**m.__dict__) for m in resp.members],
        created_at=resp.created_at,
        pinned=pinned,
        session_id=session_id,
    )


@router.get("/check-name", response_model=NameCheckOut)
async def check_name(name: str, svc: ServiceDep) -> NameCheckOut:
    available, reason = await svc.check_name(name)
    return NameCheckOut(available=available, reason=reason)


@router.post("", response_model=GroupOut, status_code=status.HTTP_201_CREATED)
async def create_group(body: GroupCreateRequest, svc: ServiceDep) -> GroupOut:
    resp = await svc.create(
        CreateGroupCommand(
            name=body.name,
            description=body.description,
            member_ids=body.member_ids,
        )
    )
    return _to_out(resp)


@router.get("", response_model=list[GroupOut])
async def list_groups(svc: ServiceDep) -> list[GroupOut]:
    """列出群组；从 Session 表附加 pinned + session_id（复用 t7 B-4-P2-CL01）。"""
    groups = await svc.list()
    pinned_map: dict[UUID, tuple[bool, UUID]] = {}
    async with session_factory() as db:
        result = await db.execute(
            select(SessionModel.id, SessionModel.group_id, SessionModel.pinned).where(
                SessionModel.type == "group"
            )
        )
        for sid, gid, pinned in result.all():
            if gid:
                pinned_map[gid] = (pinned, sid)
    return [
        _to_out(
            g,
            pinned=pinned_map.get(g.id, (False, None))[0],
            session_id=pinned_map.get(g.id, (False, None))[1],
        )
        for g in groups
    ]


@router.patch("/{group_id}", response_model=GroupOut)
async def rename_group(group_id: UUID, body: GroupRenameRequest, svc: ServiceDep) -> GroupOut:
    return _to_out(await svc.rename(RenameGroupCommand(group_id=group_id, name=body.name)))


@router.delete("/{group_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_group(group_id: UUID, svc: ServiceDep):
    await svc.delete(DeleteGroupCommand(group_id=group_id))
